vm-allocator/src/id.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 The Chromium OS Authors. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use std::fmt::{self, Display};
+use std::result;
+
+#[derive(Debug)]
+pub enum Error {
+    Overflow,
+    Duplicated,
+}
+
+impl Display for Error {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Overflow => write!(f, "Integer is overflow"),
+            Duplicated => write!(f, "Integer being allocated is duplicated"),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Manages allocating unsigned 32-bit number usage.
+/// Use `IdAllocator` whenever an unsigned 32-bit number needs to be allocated to different users.
+///
+/// # Examples
+///
+/// ```
+/// # use vm_allocator::IdAllocator;
+///   IdAllocator::new(1, std::u32::MAX).map(|mut p| {
+///       assert_eq!(p.allocate(Some(1)).unwrap(), 1);
+///       assert_eq!(p.allocate(Some(3)).unwrap(), 3);
+/// });
+/// ```
+#[derive(Debug)]
+pub struct IdAllocator {
+    start: u32,
+    end: u32,
+    used_map: Vec<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new `IdAllocator` for managing u32 usage.
+    /// * `start` - The starting number to manage.
```

Comment:
The explanation for the fields should be place in the structure before each field. `used_map` is not an argument for the new function.

Reply:
Ok, got it! Thanks for the correct comment format reminder! :)